### PR TITLE
[AIRFLOW-3046] Fix ECS Operator mistakenly reports success

### PR DESCRIPTION
**Pull request by danmactough**
Supercedes #3938 

Make sure you have checked _all_ steps below.

### Jira

- [x] My PR addresses the following [Airflow Jira](https://issues.apache.org/jira/browse/AIRFLOW-3046) issues and references them in the PR title. For example, "\[AIRFLOW-XXX\] My Airflow PR"
  - https://issues.apache.org/jira/browse/AIRFLOW-3046
  - In case you are fixing a typo in the documentation you can prepend your commit with \[AIRFLOW-XXX\], code changes always need a Jira issue.

### Description

- [x] Here are some details about my PR, including screenshots of any UI changes:

We have ECS clusters made up of EC2 spot fleets. Among other things, this means hosts can be terminated on short notice. When this happens, all tasks (and associated containers) get terminated, as well.

We expect that when that happens for Airflow task instances using the ECS Operator, those instances will be marked as failures and retried.

Instead, they are marked as successful.

As a result, the immediate downstream task fails, causing the scheduled DAG run to fail.

Here's an example of the Airflow log output when this happens:

```
[2018-09-12 01:02:02,712] {ecs_operator.py:112} INFO - ECS Task stopped, check status: {'tasks': [{'taskArn': 'arn:aws:ecs:us-east-1:111111111111:task/32d43a1d-fbc7-4659-815d-9133bde11cdc', 'clusterArn': 'arn:aws:ecs:us-east-1:111111111111:cluster/processing', 'taskDefinitionArn': 'arn:aws:ecs:us-east-1:111111111111:task-definition/foobar-testing_dataEngineering_rd:76', 'containerInstanceArn': 'arn:aws:ecs:us-east-1:111111111111:container-instance/7431f0a6-8fc5-4eff-8196-32f77d286a61', 'overrides': {'containerOverrides': [{'name': 'foobar-testing', 'command': ['./bin/generate-features.sh', '2018-09-11']}]}, 'lastStatus': 'STOPPED', 'desiredStatus': 'STOPPED', 'cpu': '4096', 'memory': '60000', 'containers': [{'containerArn': 'arn:aws:ecs:us-east-1:111111111111:container/0d5cc553-f894-4f9a-b17c-9f80f7ce8d0a', 'taskArn': 'arn:aws:ecs:us-east-1:111111111111:task/32d43a1d-fbc7-4659-815d-9133bde11cdc', 'name': 'foobar-testing', 'lastStatus': 'RUNNING', 'networkBindings': [], 'networkInterfaces': [], 'healthStatus': 'UNKNOWN'}], 'startedBy': 'Airflow', 'version': 3, 'stoppedReason': 'Host EC2 (instance i-02cf23bbd5ae26194) terminated.', 'connectivity': 'CONNECTED', 'connectivityAt': datetime.datetime(2018, 9, 12, 0, 6, 30, 245000, tzinfo=tzlocal()), 'pullStartedAt': datetime.datetime(2018, 9, 12, 0, 6, 32, 748000, tzinfo=tzlocal()), 'pullStoppedAt': datetime.datetime(2018, 9, 12, 0, 6, 59, 748000, tzinfo=tzlocal()), 'createdAt': datetime.datetime(2018, 9, 12, 0, 6, 30, 245000, tzinfo=tzlocal()), 'startedAt': datetime.datetime(2018, 9, 12, 0, 7, 0, 748000, tzinfo=tzlocal()), 'stoppingAt': datetime.datetime(2018, 9, 12, 1, 2, 0, 91000, tzinfo=tzlocal()), 'stoppedAt': datetime.datetime(2018, 9, 12, 1, 2, 0, 91000, tzinfo=tzlocal()), 'group': 'family:foobar-testing_dataEngineering_rd', 'launchType': 'EC2', 'attachments': [], 'healthStatus': 'UNKNOWN'}], 'failures': [], 'ResponseMetadata': {'RequestId': '758c791f-b627-11e8-83f7-2b76f4796ed2', 'HTTPStatusCode': 200, 'HTTPHeaders': {'server': 'Server', 'date': 'Wed, 12 Sep 2018 01:02:02 GMT', 'content-type': 'application/x-amz-json-1.1', 'content-length': '1412', 'connection': 'keep-alive', 'x-amzn-requestid': '758c791f-b627-11e8-83f7-2b76f4796ed2'}, 'RetryAttempts': 0}}
```

This PR updates the ECS Operator to include a check for this failure condition and raise an `AirflowException`.

I've also updated the `setup.py` to include `unittest2` in the `devel` section when using Python 2. The tests depend on `unittest2` being installed.

### Tests

- [x] My PR adds the following unit tests

- test_host_terminated_raises

### Commits

- [x] My commits all reference Jira issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [x] In case of new functionality, my PR adds documentation that describes how to use it.
  - When adding new operators/hooks/sensors, the autoclass documentation generation needs to be added.

### Code Quality

- [x] Passes `git diff upstream/master -u -- "*.py" | flake8 --diff`

### CODE REVIEW

File: airflow/contrib/operators/ecs_operator.py [hunk 1]
```diff
@@ -139,6 +140,10 @@ def _check_success_task(self):
             raise AirflowException(response)
 
         for task in response['tasks']:
+            if re.match(r'Host .+?\(.+?\) terminated\.', task.get('stoppedReason', '')):
```
Comment (Fokko) [inline airflow/contrib/operators/ecs_operator.py]: Maybe add a comment here to elaborate where this regex matches on. Do you know what other `stoppedReasons` might be? Just to be sure that it doesn't conflict with other reasons.

**Review by Fokko — COMMENTED** [review]
This PR makes sense to me. I'm not using Airflow on AWS myself, so I might be not the best one to judge, but wouldn't it be possible to check on the `response` that is being send back from AWS: https://github.com/apache/incubator-airflow/blob/64374000f83498f9e541d6c502f077ad26d346ff/airflow/contrib/operators/ecs_operator.py#L137

If this response is not successful, then mark the task as failed. What do you think @danmactough ?

File: airflow/contrib/operators/ecs_operator.py (hunk 1 above)
Comment (danmactough) [reply airflow/contrib/operators/ecs_operator.py]: > Do you know what other `stoppedReasons` might be?

That's a great question. The [only documentation I can find](https://docs.aws.amazon.com/AmazonECS/latest/developerguide/stopped-task-errors.html) identifies "**_some_** possible reasons and their explanations," which doesn't give a lot of confidence that the list is complete.

I would like to edit that check to match "terminated" or "stopped" though, having seen that documentation, now.

Comment (danmactough) [reply airflow/contrib/operators/ecs_operator.py]: > I would like to edit that check to match "terminated" or "stopped" though, having seen that documentation, now.

Amended with that change and added a comment as you suggested, as well, @Fokko

**Comment by danmactough** [comment]
> If this response is not successful, then mark the task as failed. What do you think @danmactough ?

I _think_ this is the only possible way, @Fokko, with the responses we get from AWS. The response we get doesn't have any failures, doesn't have a non-zero exitCode (it has no exitCode), and there's no explicit error. Here's a sample response:

```
[2018-09-12 01:02:02,712] {ecs_operator.py:112} INFO - ECS Task stopped, check status: {
  'tasks': [
    {
      'taskArn': 'arn:aws:ecs:us-east-1:111111111111:task/32d43a1d-fbc7-4659-815d-9133bde11cdc', 
      'clusterArn': 'arn:aws:ecs:us-east-1:111111111111:cluster/processing', 
      'taskDefinitionArn': 'arn:aws:ecs:us-east-1:111111111111:task-definition/foobar-testing_dataEngineering_rd:76', 
      'containerInstanceArn': 'arn:aws:ecs:us-east-1:111111111111:container-instance/7431f0a6-8fc5-4eff-8196-32f77d286a61', 
      'overrides': {
        'containerOverrides': [
          {
            'name': 'foobar-testing', 
            'command': [
              './bin/generate-features.sh', 
              '2018-09-11'
            ]
          }
        ]
      }, 
      'lastStatus': 'STOPPED', 
      'desiredStatus': 'STOPPED', 
      'cpu': '4096', 
      'memory': '60000', 
      'containers': [
        {
          'containerArn': 'arn:aws:ecs:us-east-1:111111111111:container/0d5cc553-f894-4f9a-b17c-9f80f7ce8d0a', 
          'taskArn': 'arn:aws:ecs:us-east-1:111111111111:task/32d43a1d-fbc7-4659-815d-9133bde11cdc', 
          'name': 'foobar-testing', 
          'lastStatus': 'RUNNING', 
          'networkBindings': [], 
          'networkInterfaces': [], 
          'healthStatus': 'UNKNOWN'
        }
      ], 
      'startedBy': 'Airflow', 
      'version': 3, 
      'stoppedReason': 'Host EC2 (instance i-02cf23bbd5ae26194) terminated.', 
      'connectivity': 'CONNECTED', 
      'connectivityAt': datetime.datetime(2018, 9, 12, 0, 6, 30, 245000, tzinfo=tzlocal()), 
      'pullStartedAt': datetime.datetime(2018, 9, 12, 0, 6, 32, 748000, tzinfo=tzlocal()), 
      'pullStoppedAt': datetime.datetime(2018, 9, 12, 0, 6, 59, 748000, tzinfo=tzlocal()), 
      'createdAt': datetime.datetime(2018, 9, 12, 0, 6, 30, 245000, tzinfo=tzlocal()), 
      'startedAt': datetime.datetime(2018, 9, 12, 0, 7, 0, 748000, tzinfo=tzlocal()), 
      'stoppingAt': datetime.datetime(2018, 9, 12, 1, 2, 0, 91000, tzinfo=tzlocal()), 
      'stoppedAt': datetime.datetime(2018, 9, 12, 1, 2, 0, 91000, tzinfo=tzlocal()), 
      'group': 'family:foobar-testing_dataEngineering_rd', 
      'launchType': 'EC2', 
      'attachments': [], 
      'healthStatus': 'UNKNOWN'
    }
  ], 
  'failures': [], 
  'ResponseMetadata': {
    'RequestId': '758c791f-b627-11e8-83f7-2b76f4796ed2', 
    'HTTPStatusCode': 200, 
    'HTTPHeaders': {
      'server': 'Server', 
      'date': 'Wed, 12 Sep 2018 01:02:02 GMT', 
      'content-type': 'application/x-amz-json-1.1', 
      'content-length': '1412', 
      'connection': 'keep-alive', 
      'x-amzn-requestid': '758c791f-b627-11e8-83f7-2b76f4796ed2'
    }, 
  'RetryAttempts': 0
  }
}
```

**Comment by Fokko** [comment]
Good point @danmactough. Thanks for the explanaition, like I said. I'm not an AWS guru :-) Will merge on a green CI.

File: airflow/contrib/operators/ecs_operator.py [hunk 2]
```diff
@@ -139,6 +140,15 @@ def _check_success_task(self):
             raise AirflowException(response)
 
         for task in response['tasks']:
+            # This is a `stoppedReason` that indicates a task has not
+            # successfully finished, but there is no other indication of failure
+            # in the response.
+            # See, https://docs.aws.amazon.com/AmazonECS/latest/developerguide/stopped-task-errors.html # noqa E501
+            if re.match(r'Host EC2 \(instance .+?\) (stopped|terminated)\.',
```
Comment (ashb) [inline airflow/contrib/operators/ecs_operator.py:147]: Do we need this check? A look at the doc you linked to and it sounds like any of these other stoppedReasons would also be "we had to kill your task" type things and should be an error to?

Comment (danmactough) [reply airflow/contrib/operators/ecs_operator.py:147]: > A look at the doc you linked to and it sounds like any of these other stoppedReasons would also be "we had to kill your task" type things and should be an error to?

Not quite, @ashb. "Essential container in task exited" is the stoppedReason when your task successfully finishes. 🌀

I _think_ you're probably right about some of the other stoppedReasons ("Container instance deregistration forced by user" jumps out), but the problem I'm trying to solve here is that when the host is terminated there is no other evidence of failure -- and I don't have any idea if those other stoppedReasons also have no evidence of failure or whether we can safely rely on the container's exit status. The container's exit status seems like the most precise indicator of success/failure, so I think it's preferable to use that indicator -- but in this edge case, that indicator fails.

So, you certainly may be correct that there are more edge cases, but I think we can limit this fix to the edge case we can prove.

**Comment by r39132** [comment]
@danmactough Can you rebase?

**Comment by danmactough** [comment]
@r39132 You bet. Done.

**Pull request merged**